### PR TITLE
sig-storage-lib changes to support PV Deletion protection finalizer

### PR DESCRIPTION
**Pull request by deepakkinni**
**What this PR does / why we need it**:
The fix is a sig-storage-lib-external-provisioner part of implementation for the KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2644-honor-pv-reclaim-policy

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-csi/external-provisioner/issues/546

Signed-off-by: Deepak Kinni <dkinni@vmware.com>

### CODE REVIEW

**Comment by jsafrane** [comment]
Tested together with https://github.com/kubernetes-csi/external-provisioner/pull/679

* The protection finalizer is added to existing PVs when `addFinalizer` / `HonorPVReclaimPolicy` feature gate in the provisioner is enabled.
* The protection finalizer is removed from existing PVs when the feature is turned off.
* With the feature on, the protection finalizer is removed after corresponding volume is deleted in the storage backend.
* With the feature and CSI migration on, migrated in-tree volumes are handled in the same way as CSI ones.

File: controller/controller.go [hunk 1]
```diff
@@ -1098,6 +1098,28 @@ func (ctrl *ProvisionController) syncVolume(ctx context.Context, obj interface{}
...
+	}
+
+	if modified {
+		volume.ObjectMeta.Finalizers = modifiedFinalizers
+		newVolume, err := ctrl.updatePersistentVolume(ctx, volume)
+		if err != nil {
+			return fmt.Errorf("failed to modify finalizers to %+v on volume %s err: %+v", modifiedFinalizers, volume.Name, err)
+		}
+		volume = newVolume
+	}
```
Comment (jsafrane) [inline controller/controller.go:1122]: Can you please move it to a separate function like `HandleProtectionFinalizer`? I like `syncVolume` short and self-describing.

**Review by jsafrane — COMMENTED** [review]
I have only a style comment
/approve

File: controller/controller.go (hunk 1 above)
Comment (deepakkinni) [reply controller/controller.go]: Done.

**Pull request merged**